[#5860][autodeploy] GPT-OSS MXFP4 support

Fridah-nv opened the pull request:
This PR does the following:

1. Add custom op for the routing and MoE part of `GptOssExperts` separately: 
2. initialize the model in bf16 and pattern match to the two ops above
3. Add a custom op that wraps triton_kernel's routing and MoE kernels, map the tensors from state_dict to the expected format of the triton kernels
4. Add a transform that map `torch.ops.auto_deploy.torch_moe_router` and `torch.ops.auto_deploy.torch_moe_dense_mlp` to  `torch.ops.auto_deploy.triton_mxfp4_moe`, register the new params expected from checkpoint
5. Add ep sharding for `torch.ops.auto_deploy.triton_mxfp4_moe`

env requirements for huggingface to load MXFP4 model
```
pip install triton==3.4.0
pip install kernels
pip install transformers==4.56.1
```
Currently sharding uses `tensorrt_llm._torch.modules.fused_moe.fused_moe_triton.TritonEPRouter` which requires building triton from source:
```
cd path/to/triton
pip install -r python/requirements.txt # build-time dependencies
pip install -e .
cd python/triton_kernels
pip install -e .
export PYTHONPATH=/code/tensorrt_llm/tmp/triton/python:$PYTHONPATH
```
Verified with triton version 3.5 +, commit sha: f22c53ad7

Tested on 8*H200, model is able to load onto one H200.
Tested with command:
```
python build_and_run_ad.py \
--model "/workspaces/tensorrt_llm/models/openai/gpt-oss-20b" \
--args.world-size 1 \
--args.compile-backend "torch-simple" \
--args.attn-backend "torch"

python build_and_run_ad.py \
--model "/workspaces/tensorrt_llm/models/openai/gpt-oss-20b" \
--args.world-size 8 \
--args.compile-backend "torch-simple" \
--args.attn-backend "torch"
```
Output (WS=1 and WS=8 are the same)
```
[09/23/2025-00:24:09] [TRT-LLM AUTO-DEPLOY] [I] [PROMPT 0] How big is the universe? : 1.5 trillion light years? 13.8 billion years? 13.8 billion years? 13.8 billion years? 13.8 billion years? 13.8 billion years? 13.8 billion years? 13.8 billion years? 13.8 billion years? 13.8 billion years? 13.8 billion years? 13.8 billion years? 13.8 billion years? 13.8 billion years? 13
[09/23/2025-00:24:09] [TRT-LLM AUTO-DEPLOY] [I] [PROMPT 1] In simple words and a single sentence, explain the concept of gravity: : 1) The 2nd law of Newton's law?

The second law of Newton's law states that the force acting on an object is equal to the mass of the object multiplied by its acceleration.

The second law of Newton's

The second law of Newton states that the force acting on an object is equal to the mass of the object multiplied by its acceleration.

The second law of Newton states that the force acting on an object is equal to the mass of the object multiplied by its
```
It's a bit repetitive, but the BF16 model gives something similar with torch attention backend. See discussions: https://nvidia.slack.com/archives/C08T55LHSG4/p1756162597642669?thread_ts=1756142462.162129&cid=C08T55LHSG4

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Added MXFP4-based MoE support, including fused MoE router and dense MLP operators.
  - Introduced cached attention with cache initialization and resizing in the default pipeline.
  - Export path updated to use a fused router for GPT-Oss Top-K routing.
  - Model dtype now adapts automatically based on quantization settings.

- Refactor
  - Reorganized the default deployment pipeline to include initialization/caching steps and finalize with model compilation.
  - Removed deprecated transforms (quantization, sharding, and fusion passes) from the default configuration to streamline processing.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.


- [X] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Fridah-nv committed:
pattern match and patch router and expert layer for BF16 model

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

minor update on default.yaml

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

minor update on default.yaml

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

Fridah-nv committed:
add MXFP4 mlp fused op, map to the fused op

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

add new custom op

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

add HfQuantizer

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

nvchenghaoz commented on tensorrt_llm/_torch/auto_deploy/config/default.yaml:
```diff
@@ -22,9 +22,11 @@ transforms:
+  match_dense_moe_pattern:
     stage: pattern_matcher
-  match_repeat_kv:
+  # match_moe_router_pattern:
```
Any reason for commenting out this pattern? Maybe just delete it?

nvchenghaoz commented on tensorrt_llm/_torch/auto_deploy/custom_ops/mxfp4.py:
```diff
@@ -0,0 +1,158 @@
+    #     opt_flags.update_opt_flags_constraints(constraints)
+    # # transpose the tensor so that the quantization axis is on dim1
+
+    # TODO: there is still an issue with the scales on hopper
```
Consider adding a ticket to track this TODO.

Fridah-nv committed:
sharding support

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

use TRTLLM EPRouter for sharding support

Signed-off-by: Fridah-nv <201670829+Fridah-nv@users.noreply.github.com>

Fridah-nv committed:
clean up code:handle triton import, rename file/ops to _moe, avoid code duplication

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

minor file rename

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

Fridah-nv committed:
Move HFQuantizer into QuantConfigReader system

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

update quantize_mxfp4_moe transform to read from quant config

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

Fridah-nv committed:
Move EP sharding into the sharding system

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

greg-kwasniewski1 commented on tensorrt_llm/_torch/auto_deploy/config/default.yaml:
```diff
@@ -59,6 +62,9 @@ transforms:
     stage: pattern_matcher
+  quantize_mxfp4_moe:
+    stage: pattern_matcher
+    run_shape_prop: true
```
It's more a question rather than a comment, but why does mxfp4 need shape propagation but nvfp4 does not?

Fridah-nv replied:
Please check: https://github.com/NVIDIA/TensorRT-LLM/pull/7451#discussion_r2377072321

greg-kwasniewski1 commented on tensorrt_llm/_torch/auto_deploy/custom_ops/mxfp4_moe.py:
```diff
@@ -0,0 +1,271 @@
+
+
+@torch.library.custom_op("auto_deploy::triton_mxfp4_mlp", mutates_args=())
+def triton_mxfp4_mlp(
```
Why do we need `triton_mxfp4_mlp` wrapper around `_run_mxfp4_mlp_core`? If I understand correctly, the only thing it does it packs `routing` into `_global_route_fn`. Can't it be done inside `_core`?

greg-kwasniewski1 commented on tensorrt_llm/_torch/auto_deploy/utils/sharding_utils.py:
```diff
@@ -652,6 +652,84 @@ def get_partition(lst, world_size, rank):
 
+def _slice_expert_dim(gm: GraphModule, tensor_node: Node, lo: int, hi: int) -> Node:
+    """Return tensor_node[lo:hi, ...] via aten.slice along dim 0."""
+    lo = int(lo)
```
do we need this casting to int()? It seems that `lo` and `hi`, how they are computed, they are already ints.

Fridah-nv replied:
This can be skipped, thanks for pointing out.

greg-kwasniewski1 commented on tensorrt_llm/_torch/auto_deploy/utils/sharding_utils.py:
```diff
@@ -652,6 +652,84 @@ def get_partition(lst, world_size, rank):
+        local_hi = local_lo + (base + 1)
+    else:
+        local_lo = rem * (base + 1) + (rank - rem) * base
+        local_hi = local_lo + base
```
Interesting balancing logic for the `num_experts % world_size != 0`. For me, it was hard to read what is happening,
but do we need that fine balancing? the previous EP https://github.com/NVIDIA/TensorRT-LLM/blob/0252cee4c39398dc622531e80176fb719928be75/tensorrt_llm/_torch/auto_deploy/utils/sharding_utils.py#L607
implements just:
```
local_lo = rank * ceil(num_experts, world_size)
local_hi = min(rank * ceil(num_experts, world_size), num_experts)
```
with
```
def ceil(a,b):
    return a // b + (a % b != 0)
```

greg-kwasniewski1 commented on tests/unittest/_torch/auto_deploy/unit/multigpu/custom_ops/test_mxfp4_moe_ep.py:
```diff
@@ -0,0 +1,122 @@
+from tensorrt_llm._torch.auto_deploy.distributed.common import spawn_multiprocess_job
+
+
+def _split_range_last_remainder(n: int, world_size: int, rank: int):
```
This is actually employing the load-balancing logic that i suggested earlier. I think this is incompatible with your:
```
    base = num_experts // world_size
    rem = num_experts % world_size
    if rank < rem:
        local_lo = rank * (base + 1)
        local_hi = local_lo + (base + 1)
    else:
        local_lo = rem * (base + 1) + (rank - rem) * base
        local_hi = local_lo + base
        ```
  if `rem > 1`

lucaslie commented on tensorrt_llm/_torch/auto_deploy/transform/library/mxfp4_moe.py:
```diff
@@ -0,0 +1,339 @@
+    return gm.get_parameter(name)
+
+
+def _get_submodule_of_param(gm: GraphModule, param_name: str) -> Tuple[nn.Module, str, str]:
+    # Returns (module, module_path, attr_name)
+    if "." not in param_name:
+        # param on the root
+        return gm, "", param_name
+    mod_path, _, attr = param_name.rpartition(".")
+    return gm.get_submodule(mod_path), mod_path, attr
```
I have been wanting to start a new util file for `nn.Module` utilities like `auto_deploy/utils/module.py` 

Might be a good time to start it?

lucaslie commented on tensorrt_llm/_torch/auto_deploy/transform/library/mxfp4_moe.py:
```diff
@@ -0,0 +1,339 @@
+    return gm.get_parameter(name)
+
+
+def _get_submodule_of_param(gm: GraphModule, param_name: str) -> Tuple[nn.Module, str, str]:
```
nit: the utility should work for `nn.Module` more generally, not just `GraphModule`

lucaslie commented on tensorrt_llm/_torch/auto_deploy/utils/sharding_utils.py:
@greg-kwasniewski1 , just a reminder to eventually get rid of this file and fold it into the respective `transform/library` files :)  

I'd rather have multiple sharding-related files in `transform/library` than having this util file here

greg-kwasniewski1 replied:
@lucas : do you see it just as copy-paste from `sharding_utils` to `sharding`? or do you want to make some fine-grained adjustments what goes where?

greg-kwasniewski1 reviewed (commented):
In general, looks good for me. I left comments throughout the files. My only concern is the sharding logic when num_experts % world_size != 0, and a mismatch between the test and the actual logic.

Fridah-nv commented on tensorrt_llm/_torch/auto_deploy/models/hf.py:
```diff
@@ -362,13 +365,19 @@ def _load_quantization_config(self, fetched_dir: str):
         if self._quant_config_reader is not None:
             return
-        # TODO: specified by user or auto-detect
         reader_cls = QuantConfigReaderRegistry.get("modelopt")
         result = reader_cls.from_file(fetched_dir)
+        # Fallback to HF reader if ModelOPT not present
+        if result is None:
+            hf_cls = QuantConfigReaderRegistry.get("hf")
+            try:
+                result = hf_cls.from_file(fetched_dir)
+            except Exception:
+                # Skip HF reader if it errors out during probing
+                result = None
```
Since we have two `QuantConfigReader` class now, I need a logic to auto-detect the correct class to init. The current algo is: 
if quant_config.json exists -> `ModelOPTQuantConfigReader`, 
elif quantization_config in config.json exists -> `HFQuantConfigReader` (currently only support mxfp4 as quant_method)
else no QuantConfigReader
Currently it is implemented by trying to init each `QuantConfigReader` class in order, because the `from_file` function covers the config finding process. I still think this should be in hf factory.
Another option is to let user specify the quantization backend to use. But for now since auto-detect is quite simple, I prefer to just use it, also avoid having different configs for different models, Let me know what you think :)

lucaslie replied:
I agree with you that we can have this logic :) I was just curious why you think it's better to put it in `hf.py` rather than the same file that the `QuantConfigReaderRegistry` lies in.

Feel free to ignore this comment though

Fridah-nv committed:
address reviews: update custom ops to accept both 2D and 3D input; better naming and comments use simple expert load balancing

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

minor

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

minor

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

move get_submodule to new util function

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

minor:move QuantConfigReader detection to quant_config_reader.py

Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>

Fridah-nv commented on tensorrt_llm/_torch/auto_deploy/custom_ops/mxfp4_moe.py:
```diff
@@ -0,0 +1,271 @@
+    return torch.empty_like(hidden_states)
+
+
+@torch.library.custom_op("auto_deploy::triton_mxfp4_mlp_ep", mutates_args=())
```
Renamed: `triton_mxfp4_mlp` -> `triton_mxfp4_moe` and `triton_mxfp4_mlp_ep` -> `triton_mxfp4_moe_ep`

Fridah-nv commented on tensorrt_llm/_torch/auto_deploy/utils/sharding_utils.py:
```diff
@@ -652,6 +652,84 @@ def get_partition(lst, world_size, rank):
+
+    # Prepare new args with slices for this rank
+    args = list(node.args)
+    with gm.graph.inserting_before(node):
```
This can be skipped, thanks!

Fridah-nv commented on tensorrt_llm/_torch/auto_deploy/config/default.yaml:
```diff
@@ -24,6 +24,9 @@ transforms:
     stage: pattern_matcher
+  match_dense_moe_pattern:
+    stage: pattern_matcher
+    run_shape_prop: true
```
I made an update to `default.yaml`, now one new run shape prop introduced in this PR is the one before `detect_sharding`. This is needed because the mxfp4_moe sharding detection inspects shape info in params like `gate_up_proj_bias`, and they are new params added by `quantize_mxfp4_moe`

lucaslie commented on tensorrt_llm/_torch/auto_deploy/transform/library/mxfp4_moe.py:
```diff
@@ -0,0 +1,320 @@
+
+    algo_name: str = "mxfp4"
+
+    def _apply(
```
I am curious why you decided to implement this pattern matcher without the pattern matching utility?

lucaslie reviewed (approved):
Looks good. just had two more questions

tensorrt-cicd commented:
[PR_Github #19987](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19987/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15049](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15049/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #19998](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19998/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19998](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19998/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15060](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15060/) completed with status: 'FAILURE'

Fridah-nv commented:
/bot run

tensorrt-cicd commented:
[PR_Github #20017](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20017/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15077](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15077/) completed with status: 'FAILURE'

Fridah-nv commented:
/bot run

tensorrt-cicd commented:
[PR_Github #20043](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20043/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #15099](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15099/) completed with status: 'FAILURE'

Fridah-nv committed:
Merge branch 'main' into user/fridah/oss-mxfp4

tensorrt-cicd commented:
[PR_Github #20107](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20107/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20107](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20107/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15150](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15150/) completed with status: 'FAILURE'

Fridah-nv commented:
/bot run

tensorrt-cicd commented:
[PR_Github #20114](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20114/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20114](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20114/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15157](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15157/) completed with status: 'SUCCESS'
